src/resolve/resolved-bus.c

```diff
@@ -1580,6 +1582,132 @@ static int bus_method_reset_server_features(sd_bus_message *message, void *userd
+        assert(m);
+
+        if (m->mdns_support != RESOLVE_SUPPORT_YES)
+                return sd_bus_error_set_errno(error, EPROTONOSUPPORT);
+
+        service = new0(DnssdService, 1);
+        if (!service)
+                return sd_bus_error_set_errno(error, ENOMEM);
+        service->manager = m;
+
+        r = sd_bus_message_read(message, "sssqqq", &service->name,
+                                &service->name_tpl,
+                                &service->type, &service->port,
+                                &service->priority, &service->weight);
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
+
+        s = hashmap_get(m->dnssd_services, service->name);
+        if (s)
+                return sd_bus_error_setf(error, BUS_ERROR_DNSSD_SERVICE_EXISTS, "DNS-SD service '%s' exists already", service->name);
+
+        r = dnssd_render_instance_name(service, &instance_name);
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
+
+        if (!dnssd_srv_type_is_valid(service->type))
+                return sd_bus_error_set_errno(error, EINVAL);
+
+        r = sd_bus_message_enter_container(message, SD_BUS_TYPE_ARRAY, "{ss}");
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
+
+        while ((r = sd_bus_message_enter_container(message, SD_BUS_TYPE_DICT_ENTRY, "ss")) > 0) {
+                const char *key, *value;
+                DnsTxtItem *i;
+
+                r = sd_bus_message_read(message, "ss", &key, &value);
+                if (r < 0)
+                        return sd_bus_error_set_errno(error, -r);
+
+                if (strlen_ptr(key) == 0)
+                        return sd_bus_error_set_errno(error, EINVAL);
+
+                if (!ascii_is_valid(key))
+                        return sd_bus_error_set_errno(error, EINVAL);
+
+                r = dnssd_txt_item_new(key, value, &i);
+                if (r < 0)
+                        return sd_bus_error_set_errno(error, -r);
+
+                LIST_INSERT_AFTER(items, service->txt, last, i);
+                last = i;
+        }
+
+        r = sd_bus_message_exit_container(message);
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
+
+        if (!service->txt)
+                service->txt = malloc0(offsetof(DnsTxtItem, data) + 1); /* for safety reasons we add an extra NUL byte */
+        if (!service->txt)
+                return sd_bus_error_set_errno(error, ENOMEM);
+
+        r = hashmap_ensure_allocated(&m->dnssd_services, &string_hash_ops);
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
+
+        r = hashmap_put(m->dnssd_services, service->name, service);
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
+
+        r = sd_bus_path_encode("/org/freedesktop/resolve1/dnssd", service->name, &path);
+        if (r < 0)
+                return sd_bus_error_set_errno(error, -r);
```

Comment:
i'd swap this around, move this up, so that we don't have to return a failure if we actually managed to add the entry already